docs/link-reference-definition-improvements.md

```diff
@@ -0,0 +1,88 @@
+# Link Reference Definition Improvements
+
+## Current Problems
+
+### File-by-file Insertion
+
+For the time being, if you want to get [[wiki-links]] into all files within the workspace, you'll need to generate the link reference definitions yourself file-by-file (with the assistance of Foam).
+
+### Wikilinks don't work on GitHub
+
+> **TL;DR;** [workaround](#workaround) in the end of the chapter.
+
+If you click any of the wiki-links on GitHub web UI (such as the `README.md` of a project), you'll notice that the links break with a 404 error.
+
+At the time of writing (June 28 2020) this is a known, but unsolved error. To understand why this is the case, we need to understand what we are trading off.
+
+So, why don't they work on GitHub?
+
+The three components of [[link-reference-definitions]] are link label, link destination and Link Title.
+
+The issue is the middle **link destination** component. It's configured to point to the file name **without file extension**, i.e. "file-name" instead of "file-name.md". This is to make the GitHub Pages rendering work, because if we generated the links to `file-name.md`, the links would point to the raw markdown files instead of their generated HTML versions.
+
+| Environment      | `file-name` | `file-name.md` |
+| ---------------- | ----------- | -------------- |
+| **VS Code**      | Works       | Works          |
+| **GitHub pages** | Works       | Breaks         |
+| **GitHub UI**    | Breaks      | Works          |
+
+So as you can see, we've prioritised GitHub Pages over GitHub Web UI for the time being.
+
+Ideally, we'd like a solution that works with both, but it's not defined yet (see [[link-reference-definitions]] for more details)
+
+#### Workaround
+
+For the time being, you can use relative `[markdown links](markdown-link.md)` syntax.
+
+**Pros:**
+
+- This will work on all platforms.
+
+**Cons:**
+
+- It will break the Markdown Notes [[backlinking]] support
+- Less convenient to write
+
+### Finding certain words clutter the VS Code search results
+
+Since link reference definitions have `[//begin]` and `[//end]` guards with explanatory text that use certain words, these words (like "generate") appear in VS Code search results if you happen to search matching strings from the workspace.
+
+## Improvement Proposal
+
+Problem space in essence:
+
+- During edit-time (when modifying the markdown files in an editor)
+  - link reference definitions are needed if user uses editor extensions that don't understand wikilinks
+  - link reference definitions may be annoying since they 
+    - add content to files that the user hasn't typed in by themselves
+    - get out of date if user uses a tool that doesn't autogenerate them
+    - may clutter the search results
+- During build-time (when converting markdown to html for publishing purposes)
```

Comment:
I think the build-time you describe below is not converting markdown to html, it's converting markdown to markdown (with GH Pages-friendly links) and letting GitHub pages still do the HTML generation.

Both approaches should be possible. It's not clear to me yet which should be the default for Foam:
- Have a "foam-ssg" static site generator. Powerful, publish platform agnostic, way more powerful
  - Running code means potential bugs, and if people encounter bugs in their Foam repo, they might get blocked from publishing, and have no recourse to fix it.
  - I want to build this anyway, but not sure if it should be the default
- Continue to customise the default GH Pages Jekyll/Primer with template/css/js modifications

Reply:
in general build-time means just converting the original md-files to something else. You are right that the description mostly concerned md to md transformation. Build-time as a term can certainly be somewhat confusing, especially if we consider watch-like functionality and combining that e.g. with edit-time visible changes in graphs or something like that. Hopefully in the context of this description it however brought more clarity than confusion :).